Fix nested recursion

robz opened the pull request:
When recursive functions are created inside if statements or loops, it causes several bugs in the Hack output.

If `--enable vardecl` is passed, the functions are not put into Ref->contents. If `--disable vardecl` is passed, then they are set to Ref->contents, but the Ref itself is never created.

These commits add tests that show off the bugs, and then fix the bugs by using the logic that detects when to use `->contents` in variable declarations for adding Refs. I also moved vardecl logic into php_from_rehp, since it's only needed for Hack, and because it's convenient to include it with the Ref creation logic.

robz committed:
Recursion bug in hack output

There is a bug in the Hack output when recursive functions are created inside if statements or loops.

If `--enable vardecl` is passed, the functions are not put into Ref->contents. If it is not passed, then they are set to Ref->contents, but the ref itself is never created. (In both cases, there is a bug that cause an exception at runtime). If `--noinline` is passed, it will fix the bug in a special case where the recursive function was previously being inlined, but in general this bug doesn't appear related to that flag.

robz committed:
Cleanup php_from_rehp

This fixes a lot of readability issues with this file that were making it hard to understand and debug.

robz committed:
Debug comments containing input/output vars

robz committed:
Remove topLevelIdentifiers

The sources function was relying on topLevelIdentifiers to find all variables that are declared in a function in order to figure out which variables should be refs. However, it doesn't traverse very deep into function bodies, so it caused bugs in cases where variables that need to be refs are inside if statements or loops.

Instead of using a function like topLevelIdentifiers, we can just reuse the logic for determining when to add ->contents to variable declaration in foldVars. The logic is: if the variable being declared references itself in its declaration (example: recursive functions) or if it has already been used but not declared earlier in the function body (example: mutually recursive functions) then the variable should be stored in a Ref, so the left hand side of the declaration should use ->contents.

When that it determined, the variable can be recorded in "output.refs" and then used by sources to prepend Ref declarations at the top of the function.

Notes:
- this doesn't fix recursion bugs when the vardecl flag is passed. I'm still trying to understand that case.
- there are a couple of prexisting bugs revealed by the changes to Stdlib__arg and Stdlib__genlex

robz committed:
Fix recursion for vardecl

vardecl replaces variable declarations with EBin(Eq, id, rhs) expressions (variable assignments), and then adds new None variable declarations at the top of the function.

This breaks Ref creation. Ref creation relies on detecting when a variable declaration uses itself, or when a variable being declared has already been used. Replacing declaration ASTs with assignment ASTs and adding new declarations at the top of the function prevent that detection from working.

The fix in this commit is to move the vardecl step to php_from_rehp (it's no longer run for JS). output.dec tracks mostly the same variables that vardecl moves to null declarations. It doesn't include Refs, but input vars and other variables that are declared at the top level of functions need to be removed from it (if "var x=3" is at the top of a function, vardecl will detect it and not add "var x=null").

Detecting top level variables is done in a separate pass, similar to the topLevelIdents function that was removed earlier in this stack.

The changes in rehack_tests show that this fixes the recursion bugs in the recursion tests, as well as a couple more bugs in other stdlib files. Unfortunately, the order of declarations is not retained by output.dec's StringMap, so there's a lot of extra noise.

robz committed:
Use IdentSet for vars

hhugo commented:
I'm having hard times understand the issue this PRs is trying to solve (The diff is huge). Would you mind giving a summary in the PR description ?

robz commented:
@hhugo sorry about that, I added a summary, and the intermediate commit messages are more detailed. The large size of the PR is mostly because I refactored php_from_rehp while tracking down the issues.

robz commented on rehack_tests/output/recursion/Recursion-novardecl-inline.cma.php/Recursion.php:
```diff
@@ -0,0 +1,101 @@
+    };
+    $bug = (dynamic $x) : dynamic ==> {
+      if (0 === $x) {
+        $sub->contents = (dynamic $x) : dynamic ==> {
```
bug (sub is never declared)

robz commented on rehack_tests/output/recursion/Recursion-novardecl-inline.cma.php/Recursion.php:
```diff
@@ -0,0 +1,101 @@
+    };
+    $bug__0 = (dynamic $x) : dynamic ==> {
+      if (0 === $x) {
+        $sub->contents = (dynamic $x) : dynamic ==> {
```
bug (sub is never declared)

robz commented on rehack_tests/output/recursion/Recursion-novardecl-inline.cma.php/Recursion.php:
```diff
@@ -0,0 +1,101 @@
+      $k = Vector{0, 0} as dynamic;
+      $x = 0 as dynamic;
+      for (;;) {
+        $sub->contents = (dynamic $x) : dynamic ==> {
```
bug (sub is never declared)

robz commented on rehack_tests/output/recursion/Recursion-novardecl-inline.cma.php/Recursion.php:
```diff
@@ -0,0 +1,101 @@
+      $x = 0 as dynamic;
+      for (;;) {
+        if (0 === $x) {
+          $sub->contents = (dynamic $x) : dynamic ==> {
```
bug (sub is never declared)

robz commented on rehack_tests/output/recursion/Recursion-novardecl-noinline.cma.php/Recursion.php:
```diff
@@ -0,0 +1,102 @@
+    };
+    $bug = (dynamic $x) : dynamic ==> {
+      if (0 === $x) {
+        $sub->contents = (dynamic $x) : dynamic ==> {
```
bug (sub is never declared)

robz commented on rehack_tests/output/recursion/Recursion-novardecl-noinline.cma.php/Recursion.php:
```diff
@@ -0,0 +1,102 @@
+      $k = Vector{0, 0} as dynamic;
+      $x = 0 as dynamic;
+      for (;;) {
+        $sub->contents = (dynamic $x) : dynamic ==> {
```
bug (sub is never declared)

robz commented on rehack_tests/output/recursion/Recursion-novardecl-noinline.cma.php/Recursion.php:
```diff
@@ -0,0 +1,102 @@
+      $x = 0 as dynamic;
+      for (;;) {
+        if (0 === $x) {
+          $sub->contents = (dynamic $x) : dynamic ==> {
```
bug (sub is never declared)

robz commented on rehack_tests/output/recursion/Recursion-vardecl-inline.cma.php/Recursion.php:
```diff
@@ -0,0 +1,104 @@
+    $bug = (dynamic $x) : dynamic ==> {
+      $sub = null as dynamic;
+      if (0 === $x) {
+        $sub = (dynamic $x) : dynamic ==> {return (int) (1 + $sub($x));};
```
bug (sub should be a Ref)

robz commented on rehack_tests/output/recursion/Recursion-vardecl-inline.cma.php/Recursion.php:
```diff
@@ -0,0 +1,104 @@
+    $bug__0 = (dynamic $x) : dynamic ==> {
+      $sub = null as dynamic;
+      if (0 === $x) {
+        $sub = (dynamic $x) : dynamic ==> {return (int) (1 + $sub($x));};
```
bug (sub should be a Ref)

robz commented on rehack_tests/output/recursion/Recursion-vardecl-inline.cma.php/Recursion.php:
```diff
@@ -0,0 +1,104 @@
+      $k = Vector{0, 0} as dynamic;
+      $x = 0 as dynamic;
+      for (;;) {
+        $sub = (dynamic $x) : dynamic ==> {return (int) (1 + $sub($x));};
```
bug (sub should be a Ref)

robz commented on rehack_tests/output/recursion/Recursion-vardecl-inline.cma.php/Recursion.php:
```diff
@@ -0,0 +1,104 @@
+      $x = 0 as dynamic;
+      for (;;) {
+        if (0 === $x) {
+          $sub = (dynamic $x) : dynamic ==> {return (int) (1 + $sub($x));};
```
bug (sub should be a Ref)

robz commented on rehack_tests/output/recursion/Recursion-vardecl-noinline.cma.php/Recursion.php:
```diff
@@ -0,0 +1,106 @@
+    $bug = (dynamic $x) : dynamic ==> {
+      $sub = null as dynamic;
+      if (0 === $x) {
+        $sub = (dynamic $x) : dynamic ==> {return (int) (1 + $sub($x));};
```
bug (sub should be a Ref)

robz commented on rehack_tests/output/recursion/Recursion-vardecl-noinline.cma.php/Recursion.php:
```diff
@@ -0,0 +1,106 @@
+      $k = Vector{0, 0} as dynamic;
+      $x = 0 as dynamic;
+      for (;;) {
+        $sub = (dynamic $x) : dynamic ==> {return (int) (1 + $sub($x));};
```
bug (sub should be a Ref)

robz commented on rehack_tests/output/recursion/Recursion-vardecl-noinline.cma.php/Recursion.php:
```diff
@@ -0,0 +1,106 @@
+      $x = 0 as dynamic;
+      for (;;) {
+        if (0 === $x) {
+          $sub = (dynamic $x) : dynamic ==> {return (int) (1 + $sub($x));};
```
bug (sub should be a Ref)
